src/libs/actions/Transaction.js

```diff
@@ -31,4 +52,49 @@ function addStop(transactionID, newLastIndex) {
     });
 }
 
-export {addStop, createInitialWaypoints};
+/**
+ * Saves the selected waypoint to the transaction
+ * @param {String} transactionID
+ * @param {String} index
+ * @param {Object} waypoint
+ */
+function saveWaypoint(transactionID, index, waypoint) {
+    Onyx.merge(`${ONYXKEYS.COLLECTION.TRANSACTION}${transactionID}`, {
+        comment: {
+            waypoints: {
+                [`waypoint${index}`]: waypoint,
+            },
+        },
+    });
+}
+
+function removeWaypoint(transactionID, currentIndex) {
+    // Index comes from the route params and is a string
+    const index = Number(currentIndex);
+    const transaction = lodashGet(allTransactions, transactionID, {});
+    const existingWaypoints = lodashGet(transaction, 'comment.waypoints', {});
+    const totalWaypoints = _.size(existingWaypoints);
+
+    // Prevents removing the starting or ending waypoint but clear the stored address
+    if (index === 0 || index === totalWaypoints - 1) {
+        saveWaypoint(transactionID, index, null);
+        return;
+    }
+
+    const waypointValues = _.values(existingWaypoints);
+    waypointValues.splice(index, 1);
+
+    const reIndexedWaypoints = {};
+    waypointValues.forEach((waypoint, idx) => {
+        reIndexedWaypoints[`waypoint${idx}`] = waypoint;
+    });
+
+    // Onyx.merge won't remove the null nested object values, this is a workaround
+    // to remove nested keys while also preserving other object keys
+    // Doing a deep clone of the transaction to avoid mutating the original object and running into a cache issue when using Onyx.set
+    const newTransaction = lodashCloneDeep(transaction);
+    newTransaction.comment.waypoints = reIndexedWaypoints;
```

Comment:
Instead could we spread transaction into a newTransaction object, then override only comment.waypoints?

```
const newTransaction = {
    ...transaction,
    comment: {
        ...transaction.comment,
        waypoints: {
            waypoint0: {},
            waypoint1: {},
        },
    },
};
```